### DISCUSSION TITLE
ngrx/signals, How to intialize entity state from a global store in a component store

### DISCUSSION THREAD

**Discussion by maxiuser1** [ngrx/platform#4516] [Q&A]
How to implement combine selector but in ngrx/signals, 
I have a global store that has an array, this array is fetched from a service,
In the other hand, I have a child store using entity management, I need to initialize the child store with the feched data, how can i do that?

**Comment by rainerhahnekamp** [comment]
As in the global store, you cannot initialize the store with values from a request. Instead, you would initialize the state with an empty array.

Your global store fetches the data asynchronously via an effect. In that effect, you can inject your SignalStore instance and then call the method that sets the entities.

Something like

```typescript
@Injectable()
export class GlobalStoreEffect {
  readonly #actions = inject(Actions);
  readonly #loadService = inject(LoadService);
  readonly #store = inject(MySignalStore);

  loadData$ = effect(() => { 
    returnthis.#actions.pipe(
      ofType(someAction.load), 
      switchMap(() => loadService.load()), 
      tap(data => this.#store.setData(data)), 
      map(data => someAction.loaded(data)))
    }))
}
```

**Reply by maxiuser1** [reply]
Both stores are signals, the global store is for the blogpost, and the child store is for the comments.
In the blogpost component I read the route and fetch the blogpost payload including comments, so I need to comunicate that comments to the child store "CommentsStore" that implements Entity Management, I found a solution but it requires allowSignalWrites and don't like it.

```typescript
  ngOnInit(): void {
    effect(() => {
      const blogPost = this.globalStore.blogPost();
      if (blogPost) {
        this.commentsStore.init(blogPost.comments);
      }
    }, { allowSignalWrites:true })
  }
```

**Comment by rainerhahnekamp** [comment]
Ok, but then it should be easy.

```typescript
const GlobalStore = signalStore(withState({blog: blog}))

const CommentStore = signalStore(
  withMethods(store => ({
    sync: rxMethod<Comment[]>(pipe(tap(comments => patchState(store, {comments}))
  })),
  withHooks(store => ({
    const globalStore = inject(GlobalStore);
    onInit() {
      store.sync(globalStore.blog.comments);
    }
  })
)
```

**Reply by maxiuser1** [reply]
I tried that approach yesterday but it doesn't sync when navigate to other blogpost

**Reply by rainerhahnekamp** [reply]
Both stores are globally available, i.e. you have just one instance? In that case, it has to work.

Double check that you don't call the Signal in the `sync`.

So don't `store.sync(globalStore.blog.comments());` but `store.sync(globalStore.blog.comments);`

**Reply by maxiuser1** [reply]
only global store si globally, and comments store is not globally

**Reply by rainerhahnekamp** [reply]
If you say your blogs change per route and your comment store is provided locally, there is a change it gets destroyed every time. Can you somehow provide a reproducible example?

**Reply by maxiuser1** [reply]
I created a demo here https://idx.google.com/signaldemo-1705370

**Reply by rainerhahnekamp** [reply]
Thanks, but I'm not authorized. Could give me access?

**Reply by maxiuser1** [reply]
done

**Reply by rainerhahnekamp** [reply]
You need to provide a Signal to `sync`. Otherwise it will be executed only one.

```typescript
    withHooks((store) => {
        const globalStore = inject(BlogStore);
        return {
          onInit() {
            console.log('init');
            const blogPost = globalStore.blogPost();
            if(blogPost?.comments){
                store.sync(blogPost.comments)
            }
          }
        };
      })
```

---

Try this!

```typescript
    withHooks((store) => {
        const globalStore = inject(BlogStore);
        return {
          onInit() {
            console.log('init');
            store.sync(globalStore.blogPost.comments)
          }
        };
      })
```

**Reply by maxiuser1** [reply]
Even if I pass a signal to sync the init hook is executed only one, no matter i navigate to a diferent url

```typescript
withHooks((store) => {
        const globalStore = inject(BlogStore);
        return {
          onInit() {
            store.sync(globalStore.comments);
          }
        };
    })
```

**Reply by rainerhahnekamp** [reply]
Do you have this project somewhere on a GitHub repo? IDX doesn't show me the console.

**Reply by maxiuser1** [reply]
Suire, here is https://github.com/maxiuser1/signalsdemo1

**Reply by rainerhahnekamp** [reply]
Pushed the fix as PR. I think we can close discussion here?

**Reply by maxiuser1** [reply]
sure, thanks a lot
